### PR TITLE
wallet: add config to prioritize a solution that doesn't create change in coin selection 

### PR DESCRIPTION
**Pull request by brunoerg**
Based on #22009, Bitcoin Core execute all the coin selection algorithms and decide what solution will be used based on the waste metric. However, some users may prefer a solution that doesn't create change (privacy reason). This PR adds a config option to prioritize the BnB solution in `send` and `fundrawtransaction`.

### CODE REVIEW

**Review by lsilva01 — COMMENTED** [review]
Code Review ACK d3a434d
Perhaps a functional test for this new parameter can be added.

**Comment by brunoerg** [comment]
> Code Review ACK [d3a434d](https://github.com/bitcoin/bitcoin/commit/d3a434d471da19e8e06ced10736b5efed78922d0) Perhaps a functional test for this new parameter can be added.

Functional test for it would be great, will work on it.

**Comment by prayank23** [comment]
Concept ACK

**Comment by achow101** [comment]
I don't think that this should be a startup option. The preferences for coin selection may be different per-wallet, or even on a per-transaction basis, so having it be global and apply to all wallets, all the time, seems incorrect. IMO the correct approach is to add the option to the `options` object used for `send` and `fundrawtransaction`.

**Comment by brunoerg** [comment]
> 

Thanks, @achow101.  I agree with you, the better approach would be add the option to the options object. Going to change it, thanks again.

**Comment by brunoerg** [comment]
force-pushed addressing comments from @achow101

**Comment by Rspigler** [comment]
ACK d6d634094744569d99b35bbb20f295ecdfda32bc 

I see `prioritize_no_change` bool under the `options` argument for `send` as well as `fundrawtransaction`.  Need to do more testing with spending different UTXOs.  A GUI PR would be a good followup.

**Comment by S3RK** [comment]
Changeless txs are already somewhat prioritised since creating and spending a change increases waste metric. When you force a solution without change you forego a more economically efficient solution. The question is how much would you pay for the changeless tx? Current UX hides the price of privacy from user and could lead to confusion. May a better UX would be to specify the amount you would pay for the changeless tx (if possible).

This also conflicts with possible future improvements for BnB #23367

**Comment by Rspigler** [comment]
So something like the `maxapsfee`?

How does this conflict with BnB improvements?

**Comment by S3RK** [comment]
> So something like the `maxapsfee`?

Not sure what `maxapsfee` means, but I was thinking about a config parameter that would decrease waste metric for changeless solutions by the specified amount. 

> How does this conflict with BnB improvements?

There is an idea to drop BnB upper limit. This means there will be always be a BnB solution however terrible it is. Combined with the proposed configuration parameter a user could end up with a changeless tx that drop horrendous amounts to fees.

**Comment by Rspigler** [comment]
`maxapsfee` is a config option that specifies the maximum amount of extra fees a user is willing to pay to avoid partial spends

**Comment by brunoerg** [comment]
Hi, @S3RK. Thanks for the comments, I didn't review #23367 yet, hope to do it soon.

> When you force a solution without change you forego a more economically efficient solution.
Yes, this is the purpose of this config, sometimes users don't want the more economically efficient solution, but that one that may provide a little bit more privacy (e.g. not generating change).

> Combined with the proposed configuration parameter a user could end up with a changeless tx that drop horrendous amounts to fees.

Interesting. Interesting. As I said before, I haven't checked this PR yet. But I think this config option can verify that the BnB solution is not generating horrible amounts of fees. Basically, if we have a 'decent' BnB solution, prioritize it.

File: src/dummywallet.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ void DummyWalletInit::AddWalletOptions(ArgsManager& argsman) const
         "-flushwallet",
         "-privdb",
         "-walletrejectlongchains",
-        "-unsafesqlitesync",
+        "-unsafesqlitesync"
```
Comment (achow101) [inline src/dummywallet.cpp:53]: In b6269b012fe2cad261f94480711c09ebbf47c7fa "wallet: add config to prioritize a solution that doesn't create change in coin selection"

This change is unrelated.

Comment (brunoerg) [reply src/dummywallet.cpp]: Just removed it, thanks!

**Comment by brunoerg** [comment]
Force-pushed rebasing and addressing last @achow101's comment.

File: src/wallet/coincontrol.h [hunk 2]
```diff
@@ -53,6 +53,8 @@ class CCoinControl
     bool m_avoid_partial_spends = DEFAULT_AVOIDPARTIALSPENDS;
     //! Forbids inclusion of dirty (previously used) addresses
     bool m_avoid_address_reuse = false;
+    //! Prioritize a solution that doesn't generate change when selecting coins
+    bool m_prioritize_no_change = false;
```
Comment (glozow) [inline src/wallet/coincontrol.h:57]: Should this be in `CoinSelectionParams` instead?

Comment (brunoerg) [reply src/wallet/coincontrol.h:57]: It's in both I guess

**Review by nina1234 — APPROVED** [review]

**Comment by Xekyo** [comment]
Like @S3RK, I'm a bit worried that always using the result of one coin selection algorithm could have costly outcomes. Since BnB will only use UTXOs smaller than the target, we could see results such as using 50 inputs at 100 sat/vB, while there is a single UTXO that is greater than the target.

I would prefer if this option were guarded in some fashion, e.g. that its waste must be below 3× the waste of the best solution, or the option take a parameter to define a discount for the waste of BnB solutions.

Regarding the name, I prefer avoiding negations in variable, option or parameter names. What do you think about `avoid_change` or `prefer_changeless`?

**Review by murchandamus — COMMENTED** [review]
Like @S3RK, I'm a bit worried that always using the result of one coin selection algorithm could have costly outcomes. Since BnB will only use UTXOs smaller than the target, we could see results such as using 50 inputs at 100 sat/vB, while there is a single UTXO that is greater than the target.

I would prefer if this option were guarded in some fashion, e.g. that its waste must be below 3× the waste of the best solution, or the option take a parameter to define a discount for the waste of BnB solutions.

Regarding the name, I prefer avoiding negations in variable, option or parameter names. What do you think about `avoid_change` or `prefer_changeless`?

File: src/wallet/test/coinselector_tests.cpp [hunk 3]
```diff
@@ -167,7 +167,7 @@ inline std::vector<OutputGroup>& KnapsackGroupOutputs(const std::vector<COutput>
     CoinSelectionParams coin_selection_params(/* change_output_size= */ 0,
                                               /* change_spend_size= */ 0, /* effective_feerate= */ CFeeRate(0),
                                               /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
-                                              /* tx_noinputs_size= */ 0, /* avoid_partial= */ false);
+                                              /* tx_noinputs_size= */ 0, /* avoid_partial= */ false, /* prioritize_no_change */ false);
```
Comment (luke-jr) [inline src/wallet/test/coinselector_tests.cpp:170]: as above

File: src/wallet/test/coinselector_tests.cpp [hunk 4]
```diff
@@ -717,7 +717,7 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
         CoinSelectionParams cs_params(/* change_output_size= */ 34,
                                       /* change_spend_size= */ 148, /* effective_feerate= */ CFeeRate(0),
                                       /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
-                                      /* tx_noinputs_size= */ 0, /* avoid_partial= */ false);
+                                      /* tx_noinputs_size= */ 0, /* avoid_partial= */ false, /* prioritize_no_change */ false);
```
Comment (luke-jr) [inline src/wallet/test/coinselector_tests.cpp:720]: as above

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/wallet/rpc/spend.cpp [hunk 5]
```diff
@@ -417,6 +417,7 @@ void FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& fee_out,
                 {"change_type", UniValueType(UniValue::VSTR)},
                 {"includeWatching", UniValueType(UniValue::VBOOL)},
                 {"include_watching", UniValueType(UniValue::VBOOL)},
+                {"prioritize_no_change", UniValueType(UniValue::VBOOL)},
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/rpc/spend.cpp:420]: English is not my first language but I think both are correct, aren't they?

**Comment by brunoerg** [comment]
> Regarding the name, I prefer avoiding negations in variable, option or parameter names. What do you think about `avoid_change` or `prefer_changeless`?

`avoid_change` is really good, I will adopt it. 

> Since BnB will only use UTXOs smaller than the target, we could see results such as using 50 inputs at 100 sat/vB, while there is a single UTXO that is greater than the target.

I understand your point, but if you are using `avoid_change`, it means you don't matter about the cost, you just want to avoid change even it costing much more, isn't it?

**Comment by brunoerg** [comment]
Force-pushed rebasing and addressing `avoid_change` as suggested by @Xekyo

File: src/wallet/rpc/spend.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rspigler) [reply src/wallet/rpc/spend.cpp]: `Prioritize` is the north american spelling; `prioritise` is the british spelling.  
I believe that in our docs we use the North American spelling

**Comment by Xekyo** [comment]
Some thoughts on two approaches I see here. I'll compare to the status quo and refer to the approaches as `avoid_change` and `prefer_changeless`.

**status quo**: use SelectionResult with the best waste score
- Implementation: run all coin selection algorithms, prefer SelectionResult with lowest waste
- Pro: 
  - waste score directly correlates with fee expenses for this transaction, likely produces least fees overall
  - changeless solutions are usually cheaper than solutions that create change with similarly composed input sets and get automatically preferred
- Contra:
  - Will sometimes not use an available changeless solution

**`avoid_change`**: always use changeless solution if it exists
- Implementation: run BnB first, if there is a solution, use that solution. Only run fallback if BnB didn't find a solution
- Pro:
  - probably increases overall count of changeless solutions
- Contra: 
  - will occasionally spend a chunk more in fees when the BnB solution ends up being much worse than the least wasteful solution

**`prefer_changeless`**: use changeless solution if it's not too much worse than the best solution
- Implementation: reduce waste score by a fixed or relative amount for changeless SelectionResults, e.g. reduce changeless solution's waste score by 5,000 ṩ, or reduce waste score of changeless solutions by 50%.
- Pro:
 - probably increases overall count of changeless solutions, but maybe not as much as `avoid_change`
 - limits exposure to overpaying
- Contra:
  - may skip some unattractive changeless solutions

My questions are:
- What will users think they're opting into when they activate an `avoid_change` or `prefer_changeless` option?
- How does the option change the overall cost?
- Preferring changeless solutions in general, may generally reduce the UTXO pool which actually could overall lower the frequency of changeless solutions. Would such an option actually increase or decrease changeless solutions? E.g. on #24752, increasing the allowance for creating changeless solutions surprised us in that it reduced the overall count of changeless solutions.

Given that BnB doesn't always have a solution, the wallet will sometimes need to create change anyway, how much headway would we actually make with this change?

My suggestion for a next step would be that we simulate this branch with `avoid_change` set to `true` and `false` on a few different scenarios and see how the count of changeless transactions and overall costs are affected.

**Comment by brunoerg** [comment]
Thanks, @Xekyo! 

> What will users think they're opting into when they activate an avoid_change or prefer_changeless option?

I see two scenarios: 
Using `avoid_change` users would be saying: "Please, don't generate change even if it is expensive, this is the cost I am paying for privacy". In this case, I agree `avoid_change` is so much aggressive but I think anyone using this option should be aware of the pros and cons.

Using `prefer_changeless` users would be saying: "I prefer a changeless solution but only if this one is not so expensive compared to the other ones". It's less aggressive compared to the previous one and it would have similar effects. 

Q: `Only run fallback if BnB didn't find a solution`: If user is using `avoid_change` should we run fallback if BnB didn't find a solution or should we skip it and return an error like: "It wasn't able to create a changeless solution"? Thinking about it, maybe changing the approach to `prefer_changeless` would be better. It's not a good experience for the user to have an `avoid_change` option which can generate a solution with change (users don't understand our algorithms and can open an issue relating a bug, for example).

-------
> My suggestion for a next step would be that we simulate this branch with avoid_change set to true and false on a few different scenarios and see how the count of changeless transactions and overall costs are affected.

Yeah, great, I am working on it!

**Comment by Xekyo** [comment]
That's a good point about not running other algorithms at all when BnB doesn't return a solution. I think that a `prefer_changeless` option would be less of a footgun and would have a better UX. Perhaps `avoid_change` would work better in a flow where the user reviews the final transaction before sending it.

File: src/wallet/spend.cpp [hunk 6]
```diff
@@ -384,6 +384,10 @@ std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAm
     // Note that unlike KnapsackSolver, we do not include the fee for creating a change output as BnB will not create a change output.
     std::vector<OutputGroup> positive_groups = GroupOutputs(wallet, coins, coin_selection_params, eligibility_filter, true /* positive_only */);
     if (auto bnb_result{SelectCoinsBnB(positive_groups, nTargetValue, coin_selection_params.m_cost_of_change)}) {
+        bnb_result->ComputeAndSetWaste(CAmount(0));
```
Comment (ishaanam) [inline src/wallet/spend.cpp:387]: Is this line necessary? I think SelectCoinsBnB() already calls ComputeAndSetWaste() for this result.

File: src/wallet/rpc/spend.cpp [hunk 7]
```diff
@@ -1241,6 +1248,11 @@ RPCHelpMan send()
             // be overridden by options.add_inputs.
             coin_control.m_add_inputs = rawTx.vin.size() == 0;
             SetOptionsInputWeights(options["inputs"], options);
+
+            if (options.exists("avoid_change")) {
```
Comment (ishaanam) [inline src/wallet/rpc/spend.cpp:1252]: I think this can be removed from send() because it is already run in FundTransaction().

**Review by ishaanam — COMMENTED** [review]
I know that the author of this PR may implement `prefer_changeless` instead of `avoid_change`, but I think that my comments are applicable under either option.

File: src/wallet/rpc/spend.cpp (hunk 7 above)
Comment (brunoerg) [reply src/wallet/rpc/spend.cpp:1252]: Hmm, interesting! Gonna check it

**Comment by brunoerg** [comment]
Closing for now as I haven't been working on it for a while
